Add `Segmented Group` documentation

jorytindall opened the pull request:
### :pushpin: Summary

This is the first pass at documentation for the ~~`Input Group`~~ `Segmented Group`, containing:
* Overview, description of the component
* Guidelines detailing the `Base` component, `Segment` primitives, and examples for how they can be assembled
* Anatomy of the base component
* States of the `Segments`
* Addition of the preview image

The content is currently nested under the form components category, which I believe we agreed upon initially, and I believe still makes sense given the makeup of the primitives and the use cases.

I don't expect to have this visible to the public yet (prior to the component development) so given the necessary approvals I will hide it. Though, it would be interesting to consider marking this with an experimental badge at the top just to differentiate it from released components. Thoughts?

### :link: External links

<!-- Issues, RFC, etc. -->
Jira ticket: [HDS-899](https://hashicorp.atlassian.net/browse/HDS-899)
Figma file: [Input Group](https://www.figma.com/file/noyY6dUMDYjmySpHcMjhkN/HDS-Product---Components?node-id=30582%3A48492&t=Mlaxk6zAkUw8iQHl-1)


[HDS-899]: https://hashicorp.atlassian.net/browse/HDS-899?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

didoo commented on website/docs/components/form/input-group/partials/guidelines/guidelines.md:
```diff
@@ -0,0 +1,114 @@
+
+!!! Info
+
+The way in which these components are published and assembled together are fundamenally different between the Figma Ember components.
```
"between the Figma Ember components." ??

jorytindall replied:
Should be "Figma and Ember compoennts"

didoo commented on website/docs/components/form/input-group/partials/guidelines/guidelines.md:
```diff
@@ -0,0 +1,114 @@
+### Button types
+
+- **Default**: a default button can perform a function or interaction within an Input Group, e.g., generating an object, string, or value in a field it is connected to. This is an extension of the Helios [Button](/components/button).
+- **Toggle**: a toggle button can be used to filter or narrow the available parameters when searching within a data set or group of objects. This is an extension of the Helios [Toggle Button](/components/toggle-button).
```
there's no such link as `toggle-button`

alex-ju replied:
Indeed, this doesn't match what we have in code. There is a `<Hds::Dropdown::Toggle::Button />`, but it's not of much use (functionally) unless it's part of the dropdown component `<Hds::Dropdown as |dd|><dd.ToggleButton /></Hds::Dropdown>`. So I guess, from a technical point of view, it probably makes sense to split the two types as `<Hds::Button>` and `<Hds::Dropdown>`. I can give this a bit more thinking 🤔

jorytindall replied:
We refer to this as a `ToggleButton` in Figma, probably should discuss a bit more how much the semantics should overlap here.

didoo reviewed (commented):
added a couple of quick comments

alex-ju commented on website/docs/components/form/input-group/partials/guidelines/overview.md:
```diff
@@ -0,0 +1 @@
+An Input Group combines one or more input fields, buttons, toggles, and other form controls into a single component to handle multi-directional data, search, and fine-grained filtering within an object or dataset.
```
I find it difficult to process what 'multi-directional data' refers to – would you be able to expand on this a bit?

alex-ju reviewed (commented):
Thank you for putting this together, Jory! 🙌  Certainly not the most straightforward component to document 😁  I left a few comment and suggestions, but overall looks good so far.

heatherlarsen commented on website/docs/components/form/input-group/partials/guidelines/guidelines.md:
```diff
@@ -0,0 +1,114 @@
+
+The content with the Input Group is dependent on the context:
+
+- If used a within a form, the Input Group should be used within a [Form::Fieldset](/components/form/primitives) primitive and requires a single label.
```
```suggestion
- If used in a form, the Input Group should be used within a [Form::Fieldset](/components/form/primitives) primitive and requires a single label.
```

heatherlarsen commented on website/docs/components/form/input-group/partials/specifications/anatomy.md:
```diff
@@ -0,0 +1,26 @@
+
+### Base component
+
+![Input Group Base Anatomy](/assets/components/form/input-group/input-group-base-anatomy.png =800x*)
```
This image appears to be stretched and grainy. I think the size declaration is probably a little off.

heatherlarsen commented on website/docs/components/form/input-group/partials/specifications/anatomy.md:
```diff
@@ -0,0 +1,26 @@
+
+### Subcomponent anatomy
+
+The subcomponents contained within an Input Group are extensions of Helios form components and have the same anatomy as the master components they map to.
```
We try to avoid using "master" within HashiCorp because of the negative connotation. Usually, I see "main" used instead.

```suggestion
The subcomponents contained within an Input Group are extensions of Helios form components and have the same anatomy as the main components they map to.
```

jorytindall replied:
Good catch, thanks for getting that

heatherlarsen commented on website/docs/components/form/input-group/partials/specifications/anatomy.md:
```diff
@@ -0,0 +1,26 @@
+
+| Element | Usage |
+|---------|-------|
+| Input | Required; can be one of many types extended via the [Text Input](/components/form/text-input) component. |
```
There was an example on the "Guidelines" page that didn't include an Input; it only included Buttons/Toggles, so you may want to reconsider how this is documented.

jorytindall replied:
Within the base component this would be required, but using the primitives could create any type of custom combination. I'll look into adding a bit more clarity around the assembly of subcomponents and how the base component is related/builds upon.

heatherlarsen commented on website/docs/components/form/input-group/partials/specifications/anatomy.md:
```diff
@@ -0,0 +1,26 @@
+
+## States
+
+The Input Group doesn't have it's own state intrinsically, instead, the nested component primitives should maintain their own interactive state corresponding with the master component specs.
```
```suggestion
The Input Group doesn't have its own state intrinsically; instead, the nested component primitives should maintain their own interactive state corresponding with the main component specs.
```

heatherlarsen reviewed (commented):
Mostly copy edits, but made a few suggestions as well. Won't approve/request changes so as not to block progress while I'm out.

jorytindall commented on website/docs/components/form/input-group/partials/guidelines/guidelines.md:
```diff
@@ -0,0 +1,114 @@
+- When filtering complex data, searching within a set of filters, and creating complex filtering patterns.
+- When collecting information in a form that benefits from a "connected" visual representation, e.g., a key/value pair.
+- When an action performs a generative function rather than a submit function, e.g., generating API credentials.
+- As a complex data input, e.g., uploading a document or connecting to an external resource.
```
That's a bit of the rub with this component, visually an upload could use the primitives defined in Figma, but the component in Ember would use a different button/input type (check me on this @alex-ju, not 100% sure on the exact implementation). I'll remove this point for now though.

alex-ju replied:
Sorry for the later reply on this – it slipped through. Indeed, I see that in Figma it can be achieved using this component but in code, we have a semantic element that generates both the button and the 'field' and we'd like want to make use of it, for usability and accessibility considerations (can expand on this).

MelSumner reviewed (approved):
It seems like you've addressed the feedback, and it looks good to me! 👍

KristinLBradley commented on website/docs/components/form/input-group/partials/guidelines/guidelines.md:
```diff
@@ -0,0 +1,115 @@
+
+!!! Dont
+
+Don't use an Input Group when fields are related, but aren't stored in the same object upon submission. This example features a payment form with an Input Group and label, but uses a placeholder instead of a label to communicate what information the field collects.
```
```suggestion
Don’t use an Input Group when fields are related, but aren’t stored in the same object upon submission. This example features a payment form with an Input Group and label, but uses a placeholder instead of a label to communicate what information the field collects.
```

KristinLBradley commented on website/docs/components/form/input-group/partials/specifications/anatomy.md:
```diff
@@ -0,0 +1,26 @@
+
+!!! Info
+
+The Input Group `Base` component is only available in Figma and is best used in less complex scenarios which don't perform complex filtering or functions.
```
```suggestion
The Input Group `Base` component is only available in Figma and is best used in less complex scenarios which don’t perform complex filtering or functions.
```

KristinLBradley commented on website/docs/components/form/input-group/partials/specifications/anatomy.md:
```diff
@@ -0,0 +1,26 @@
+
+## States
+
+The Input Group doesn't have its own state intrinsically; instead, the nested component primitives should maintain their own interactive state corresponding with the main component specs.
```
```suggestion
The Input Group doesn’t have its own state intrinsically; instead, the nested component primitives should maintain their own interactive state corresponding with the main component specs.
```

alex-ju commented on website/docs/components/form/segmented-group/partials/guidelines/overview.md:
```diff
@@ -0,0 +1 @@
+A Segmented Group combines two or more elements (`Segments`) into a single component to manage complex data, search, and fine-grained filtering within an object or dataset.
```
❤️

didoo replied:
```suggestion
A Segmented Group combines two or more elements (Segments) into a single component to manage complex data, search, and fine-grained filtering within an object or dataset.
```

alex-ju commented on website/docs/components/form/segmented-group/partials/guidelines/guidelines.md:
```diff
@@ -0,0 +1,116 @@
+
+!!! Info
+
+The way in which these components are published and assembled is fundamentally different between the Figma and Ember components.
```
```suggestion
How these components are published and assembled is fundamentally different between the Figma and Ember components.
```

alex-ju commented on website/docs/components/form/segmented-group/partials/guidelines/guidelines.md:
```diff
@@ -0,0 +1,116 @@
+
+### When not to use
+
+- To submit a form or set of inputs. A Segmented Group can be used within a form to collect complex data, but should not submit that data via a form method. Use the guidelines in [Form patterns](/patterns/form-patterns) instead.
```
🤔 hmm, not sure about this – why would we not want them to submit the data using a standard form? happy to catch-up on this

alex-ju replied:
ah, I see! yes, that sounds like valuable guidance – may not require rewording, it was probably just me reading through the lines 😅

alex-ju reviewed (approved):
Looking great! Only have one question about a recommendation otherwise it looks correct to me

KristinLBradley commented on website/docs/components/form/segmented-group/partials/specifications/anatomy.md:
```diff
@@ -0,0 +1,35 @@
+
+!!! Info
+
+The Segmented Group `Base` component is only available in Figma and is suited for less complex scenarios which don't perform complex filtering or functions.
```
```suggestion
The Segmented Group `Base` component is only available in Figma and is suited for less complex scenarios which don’t perform complex filtering or functions.
```

didoo commented on website/docs/components/form/segmented-group/index.md:
```diff
@@ -0,0 +1,26 @@
+  figma: https://www.figma.com/file/noyY6dUMDYjmySpHcMjhkN/HDS-Product---Components?node-id=30150%3A45701&t=akpWrhSdwygOH6md-1
+previewImage: assets/illustrations/components/form/segmented-group.jpg
+navigation:
+  keywords: [ 'filter', 'connected', 'fields', 'set' ]
```
do we want to add more keywords here, related to what it is (eg. `group`, `segment`) and contains (eg. `button`, `dropdown`, `input`, `select`)?

jorytindall replied:
Added these, the only thing that gives me a little pause is including the items it can contain (might muddy up the search a little bit). I did a bit of testing, it seems reasonable but if a consumer doesn't have any concept of a what a `Segmented Group` is they likely won't understand the correlation right off the bat. This probably isn't a problem though.

didoo commented on website/docs/components/form/segmented-group/partials/guidelines/guidelines.md:
```diff
@@ -0,0 +1,116 @@
+## Usage
+
+A Segmented Group is an agnostic container that wraps two or more `Segments`, making the relevant use cases broad.
```
No need of the "code" visual treatment: that's not the name of a component.

```suggestion
A Segmented Group is an agnostic container that wraps two or more Segments, making the relevant use cases broad.
```

didoo commented on website/docs/components/form/segmented-group/partials/guidelines/guidelines.md:
```diff
@@ -0,0 +1,116 @@
+How these components are published and assembled is fundamentally different between the Figma and Ember components.
+
+- In Figma, we publish multiple primitive `Segment` components (`Segmented Button`, `Segmented Toggle`, and `Segmented Input`) that are intended to be assembled in an auto layout container. A pre-assembled `Base` component is also made available for more simple instances.
+- The Ember component is a container that enforces consistent styling and grouping of nested Helios form components. It does not manage any logic or functionality out of the box, these are left up to the consumer to implement based on the nested components.
```
```suggestion
- The Ember component is a container that enforces consistent styling and grouping of nested Helios input fields and action components. It does not manage any logic or functionality out of the box, these are left up to the consumer to implement based on the nested components.
```

jorytindall replied:
Ah good catch there 👏

didoo commented on website/docs/components/form/segmented-group/partials/guidelines/guidelines.md:
```diff
@@ -0,0 +1,116 @@
+
+## Segment components
+
+A Segmented Group can consist of multiple primitive Segments that vary in their properties and intended input value. We recommend limiting usage in Figma to the defined `Segmented Button`, `Segmented Toggle` and `Segmented Input` primitives to ensure consistent styling and grouping. The Ember component supports a generic element in either the leading or trailing position if a custom solution is r ...
```
See comment below about "Toggle" vs "Dropdown" (and check in other parts of the documentation too)

```suggestion
A Segmented Group can consist of multiple primitive Segments that vary in their properties and intended input value. We recommend limiting usage in Figma to the defined `Segmented Button`, `Segmented Dropdown` and `Segmented Input` primitives to ensure consistent styling and grouping. The Ember component supports a generic element in either the leading or trailing position if a custom solution is required.
```

didoo replied:
@alex-ju the statement "The Ember component supports a generic element in either the leading or trailing position if a custom solution is required" I think is not accurate: since it's a yielded component, it can be in any position.

alex-ju replied:
Indeed, it can be anywhere, leading, trailing, or anywhere in the middle.

didoo commented on website/docs/components/form/segmented-group/partials/guidelines/guidelines.md:
```diff
@@ -0,0 +1,116 @@
+
+A `Segmented Button` can perform a function or interaction within a Segmented Group, e.g., generating an object, string, or value in a field it is connected to. It is an extension of the Helios [Button](/components/button).
+
+### Segmented Toggle
```
[blocker] What it's called here "Toggle" is technically a "Dropdown", we should use the correct term otherwise it can be confusing (a form Toggle is [something different](https://helios.hashicorp.design/components/form/toggle))

(check also in other parts of the documentation, and Figma too I think)

didoo commented on website/docs/components/form/segmented-group/partials/guidelines/guidelines.md:
```diff
@@ -0,0 +1,116 @@
+
+### Segmented Toggle
+
+A `Segmented Toggle` can be used to filter or narrow the available parameters when searching within a dataset or group of objects. This is an extension of the Helios [Toggle Button](/components/dropdown#toggle) and is often used in the context of a [Dropdown](/components/dropdown).
```
Unless I am missing something, this is not factually correct: it's a Dropdown, not just an extension of the Dropdown toggle. Right?

didoo commented on website/docs/components/form/segmented-group/index.md:
```diff
@@ -0,0 +1,26 @@
+  @include "partials/specifications/anatomy.md"
+</section>
+
+<section data-tab="Accessibility">
```
the "Accessibility" section is empty

alex-ju replied:
I will raise a PR against this one with code docs and accessibility

didoo reviewed (changes requested):
@jorytindall @alex-ju I added a few comments, one of which is a blocker.

On a more general note, I think we need to make a call if this component lives under `forms` or not: in this PR yes, in the code implementation and showcase it doesn't.
Considered how it's implemented, and how it has become, personally I think it should _not_ live under the "form" folder.

alex-ju commented:
> @jorytindall @alex-ju I added a few comments, one of which is a blocker.
> 
> On a more general note, I think we need to make a call if this component lives under `forms` or not: in this PR yes, in the code implementation and showcase it doesn't. Considered how it's implemented, and how it has become, personally I think it should _not_ live under the "form" folder.

It totally slipped my mind during the review, but I ran into it this morning. Indeed, there is a misalignment here – I'll wait for Jory's thoughts on this, but since it covers filtering and 'form' usage, I lean towards having it at the root level and not under forms.

jorytindall commented:
Couple of updates based on feedback:
* I've updated the hierarchy of the docs to exist at the root level as opposed to within forms, @alex-ju I agree, this makes sense given the update in strategy
* I've updated all references of `Toggle` to `Dropdown` to match the name of the component.
* I've added the GitHub link.

Given these updates, and @alex-ju 's PR for [code docs](https://github.com/hashicorp/design-system/pull/1332) it seems like this is ready to go!

didoo commented on website/docs/components/segmented-group/partials/specifications/anatomy.md:
```diff
@@ -0,0 +1,35 @@
+| Element | Usage |
+|---------|-------|
+| Segmented Input | Required; can be one of many types which extend the [Text Input](/components/form/text-input) component. |
+| Segmented Button | Required; can be either `button` or `toggle` which are extensions of the Helios [Button](/components/button) and [Toggle Button](/components/dropdown#toggle) components. |
```
```suggestion
| Segmented Button | Required; can be either `button` or `dropdown` which are extensions of the Helios [Button](/components/button) and [Dropdown](/components/dropdown) components. |
```

didoo commented on website/docs/components/segmented-group/partials/specifications/anatomy.md:
```diff
@@ -0,0 +1,35 @@
+The Segments contained within a Segmented Group are extensions of Helios form components and have the same anatomy as the main components they extend. For more information on the specifications of each Segment, refer to their respective documentation:
+
+- [Button](/components/button)
+- [Toggle](/components/dropdown#toggle)
```
```suggestion
- [Dropdown](/components/dropdown)
```

didoo commented on website/docs/components/segmented-group/partials/guidelines/guidelines.md:
```diff
@@ -0,0 +1,116 @@
+## Related
+
+- [Button](/components/button)
+- [Dropdown Button](/components/dropdown-button)
```
```suggestion
- [Dropdown](/components/dropdown)
```

didoo reviewed (approved):
@jorytindall I'll approve to unblock, but notice that there are some leftovers of `Toggle` (and one URL was not correct)
